saleor/order/tasks.py

```diff
@@ -21,3 +38,81 @@ def send_order_updated(order_ids):
     manager = get_plugins_manager()
     for order in Order.objects.filter(id__in=order_ids):
         manager.order_updated(order)
+
+
+def _batch_ids(iterable, batch_size=1):
+    length = len(iterable)
+    for index in range(0, length, batch_size):
+        yield iterable[index : min(index + batch_size, length)]
+
+
+def _queryset_in_batches(queryset):
+    ids = queryset.values_list("id", flat=True)
+    for ids_batch in _batch_ids(ids, ORDER_BATCH_SIZE):
+        yield ids_batch
+
+
+def _bulk_release_voucher_usage(order_ids):
+    voucher_orders = Order.objects.filter(
+        voucher=OuterRef("pk"),
+        id__in=order_ids,
+    )
+    count_orders = voucher_orders.annotate(
+        count=Func(F("pk"), function="Count")
+    ).values("count")
+
+    Voucher.objects.filter(
+        Exists(voucher_orders),
+        usage_limit__isnull=False,
+    ).annotate(
+        order_count=Subquery(count_orders)
+    ).update(used=F("used") - F("order_count"))
+
+    voucher_customer_orders = Order.objects.filter(
+        voucher=OuterRef("voucher__id"),
+        user_email=OuterRef("customer_email"),
+        id__in=order_ids,
+    )
+
+    VoucherCustomer.objects.filter(Exists(voucher_customer_orders)).delete()
+
+
+def _call_expired_order_events(order_ids, manager):
+    orders = Order.objects.filter(id__in=order_ids)
+    for order in orders:
+        call_event(manager.order_expired, order)
+        call_event(manager.order_updated, order)
+
+
+def _expire_orders(manager, now):
+    qs = Order.objects.filter(
+        ~Exists(TransactionItem.objects.filter(order=OuterRef("pk"))),
+        ~Exists(Payment.objects.filter(order=OuterRef("pk"))),
+        status=OrderStatus.UNCONFIRMED,
+        channel__expire_orders_after__isnull=False,
+        channel__expire_orders_after__gt=0,
+        channel__expire_orders_after__lte=Extract(now - F("created_at"), "epoch"),
+    )
+    for ids_batch in _queryset_in_batches(qs):
+        with traced_atomic_transaction():
+            Order.objects.filter(id__in=ids_batch).update(status=OrderStatus.EXPIRED)
+
+            _bulk_release_voucher_usage(ids_batch)
+            deallocate_stock_for_orders(ids_batch, manager)
+            _call_expired_order_events(ids_batch, manager)
+
+
+@app.task(soft_time_limit=60 * 30)
+def expire_orders_task():
+    now = timezone.now()
+    task_name = "expire_orders"
+    manager = get_plugins_manager()
+    try:
+        with celery_task_lock(task_name) as (lock_obj, acquired):
+            if lock_obj.created_at < now - timezone.timedelta(hours=1):
```

Comment:
Is there a case when `acquired` is `True` and this here will be also `True`?

Reply:
I think that's such case should not happen, but added that logging just in case. Do you think this check unnecessary?

Reply:
Logging is fine, I'm thinking of case where both `if` would return `True`. `CeleryTask` has `unique=True` on task name so there can't be race condition case where there would be 2 records. I think I would be ok to just do
```suggestion
            if not acquired:
                if lock_obj.created_at < now - timezone.timedelta(hours=1):
                    logger.error("%s task exceeded 1h working time.", [task_name])
            else:
                _expire_orders(manager, now)
```